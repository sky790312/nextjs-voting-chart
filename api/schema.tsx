export interface PoolSchema {
  answer: AnswerSchema
  id: number
  publishedDate: number
  title: string
}

export interface AnswerSchema {
  type: 'Single' | 'Multi'
  options: {
    id: number
    label: string
  }[]
}
