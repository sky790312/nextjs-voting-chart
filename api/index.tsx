import * as fakeData from '@/api/poll.json'

export const fetchPolls = async () => {
  const data = (await (fakeData as any)?.default) ?? {}
  return data
}

export const fetchPoll = async (id: number) => {
  const data = (await (fakeData as any)?.default) ?? {}
  const poll = data.polls.find(poll => poll.id === id)
  return poll
}
