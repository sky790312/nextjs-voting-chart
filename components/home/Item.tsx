/* eslint-disable @typescript-eslint/camelcase */
import styled from 'styled-components'
import { PoolSchema } from '@/api/schema'
import dayjs from 'dayjs'
import { useRouter } from 'next/router'
import { FlexCenterContainer } from '@/GlobalStyles'

interface Props {
  item: PoolSchema
}

export const Item: React.FC<Props> = ({ item }) => {
  const router = useRouter()

  const goDetail = id =>
    router.push({
      pathname: '/detail',
      query: { id },
    })

  return (
    <ItemContainer onClick={() => goDetail(item.id)}>
      <div>img</div>
      <div>
        <h3>{dayjs(item.publishedDate).format('ll')}</h3>
        <p>{item.title}</p>
      </div>
    </ItemContainer>
  )
}

const ItemContainer = styled(FlexCenterContainer)`
  width: 100%;
  display: flex;
  padding: 20px;
  border-bottom: 1px solid ${props => props.theme.colors.gray};
  cursor: pointer;
`
