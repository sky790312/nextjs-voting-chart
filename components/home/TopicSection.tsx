import React from 'react'
import { PoolSchema } from '@/api/schema'
import { getDonutChartData } from '@/utils'
import styled from 'styled-components'
import dayjs from 'dayjs'
import { Doughnut } from 'react-chartjs-2'

interface Props {
  poll: PoolSchema
}

export const TopicSection: React.FC<Props> = React.memo(({ poll }) => {
  const donutChartConfig = {
    labels: poll?.answer.options.map(option => option.label),
    data: poll?.answer.options.map((option, index) => (index + 1) * 10),
  }
  const data = getDonutChartData(donutChartConfig)

  const handleAnswer = e => {
    console.log(e.target.value)
  }

  return (
    <TopicSectionContainer>
      <div>
        <h3>Today&apos;s Poll</h3>
        <p>
          {poll?.title}
          <span>{dayjs(poll?.publishedDate).format('ll')}</span>
        </p>
        <div>
          {poll?.answer.options.map(option => (
            <span key={option.id}>
              <input
                type={poll?.answer.type === 'Single' ? 'radio' : 'checkbox'}
                id={String(option.id)}
                value={option.label}
                name={String(poll?.id)}
                onChange={handleAnswer}
              />
              <label htmlFor={String(option.id)}>{option.label}</label>
            </span>
          ))}
        </div>
        <p>total number of votes recorded: </p>
      </div>
      <div>
        <Doughnut data={data} />
      </div>
    </TopicSectionContainer>
  )
})

const TopicSectionContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 20px;
  background-color: ${props => props.theme.colors.gray};

  @media (min-width: 1140px) {
    flex-direction: row;

    > div {
      width: 50%;
    }
  }
`
