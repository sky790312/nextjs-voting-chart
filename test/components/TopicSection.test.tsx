import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { TopicSection } from '@/components/home/TopicSection'
import { PoolSchema } from '@/api/schema'

describe('TopicSection component', () => {
  it('matches snapshot', () => {
    const item: PoolSchema = {
      id: 0,
      title: 'test',
      publishedDate: 123123,
      answer: {
        type: 'Single',
        options: [
          {
            id: 1,
            label: 'test answer',
          },
        ],
      },
    }
    const { asFragment } = render(
      withProviders(<TopicSection poll={item} />),
      {},
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
