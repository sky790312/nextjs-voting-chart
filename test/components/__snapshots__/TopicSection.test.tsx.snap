// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicSection component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TopicSection__TopicSectionContainer-m3h1n7-0 dlphWy"
  >
    <div>
      <h3>
        Today's Poll
      </h3>
      <p>
        test
        <span>
          ll
        </span>
      </p>
      <div>
        <span>
          <input
            id="1"
            name="0"
            type="radio"
            value="test answer"
          />
          <label
            for="1"
          >
            test answer
          </label>
        </span>
      </div>
      <p>
        total number of votes recorded: 
      </p>
    </div>
    <div>
      <canvas
        height="150"
        width="300"
      />
    </div>
  </div>
</DocumentFragment>
`;
